src/wit_gen.rs

```diff
@@ -121,6 +122,84 @@ impl<F: Default, TraceArgs> TraceGenerator<F, TraceArgs> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct AutoTraceGenerator<F> {
+    auto_signals: HashMap<UUID, HashMap<Queriable<F>, ASTExpr<F>>>,
+}
+
+impl<F> Default for AutoTraceGenerator<F> {
+    fn default() -> Self {
+        Self {
+            auto_signals: Default::default(),
+        }
+    }
+}
+
+impl<F: Clone, TraceArgs> From<&Circuit<F, TraceArgs>> for AutoTraceGenerator<F> {
+    fn from(circuit: &Circuit<F, TraceArgs>) -> Self {
+        let auto_signals = circuit
+            .step_types
+            .iter()
+            .map(|(&uuid, step_type)| (uuid, (*step_type).auto_signals.clone()))
+            .collect();
+
+        Self { auto_signals }
+    }
+}
+
+impl<F: Field + Eq + PartialEq + Hash + Clone> AutoTraceGenerator<F> {
+    pub fn generate(&self, mut witness: TraceWitness<F>) -> TraceWitness<F> {
+        for step_instance in witness.step_instances.iter_mut() {
+            let uuid = step_instance.step_type_uuid;
+
+            if let Some(auto_signals) = self.auto_signals.get(&uuid) {
+                self.step_gen(auto_signals, step_instance)
+            }
+        }
+
+        witness
+    }
+
+    fn step_gen(
+        &self,
+        auto_signals: &HashMap<Queriable<F>, ASTExpr<F>>,
+        witness: &mut StepInstance<F>,
+    ) {
+        let mut pending = auto_signals
+            .keys()
+            .filter(|s| !witness.assignments.get(s).is_some())
+            .map(|s| s.clone())
+            .collect::<Vec<Queriable<F>>>()
+            .clone();
+
+        let mut pending_amount = pending.len();
+
+        while pending_amount > 0 {
+            pending = pending
+                .clone()
+                .into_iter()
+                .filter(|s| {
+                    auto_signals
+                        .get(&s)
+                        .expect("auto definition not found")
+                        .eval(&witness.assignments)
+                        .map(|value| witness.assign(*s, value));
+
+                    !witness.assignments.get(s).is_some()
+                })
+                .map(|s| s.clone())
+                .collect::<Vec<Queriable<F>>>()
+                .clone();
+
+            // in each round at least one new signal should be assigned
+            if pending.len() == pending_amount {
+                panic!("cannot infer some auto signals")
+            }
+            pending_amount = pending.len()
+        }
```

Comment:
Is my understanding correct that each while loop attempts to assign as many pending signals as possible given the existing assignments? Then in the next loop it assigns additional signals that might depend on the assignment of the prior loop?

Is there a way to make the `panic!` message more specific that shows which signal(s) cannot be inferred?

Reply:
Yes, but I was leaving that for a small task for new people